### DISCUSSION TITLE
Check if a word or words are already highilighted

### DISCUSSION THREAD

**Discussion by mameIIas** [pymupdf/PyMuPDF#1665] [Looking for help] [answered]
Is there a way to know if a word is highlighted?
I'm trying to anonimize a pdf, so first I replace the origin word for a string of v like "vvvvv vvv vvvv". And then run the code below, which looks for that v together and highlight them, but as it is iterating it highlights over and over the same part.
This code looks for that v highlight in a color and replace the word color for white.
So can I know if a part is highlighted like:
if nombre.highlight == True:
     pass
else:
code below

```
for page in doc:
	for i in range(21, 1, -1):
		nombre = "v"*i
		for r in page.search_for(nombre):
			a=page.add_highlight_annot(r)
			a.setColors({"stroke":color})
			a.update()
			for b in page.getText("dict", clip=r)["blocks"]:
				for l in b["lines"]:
					for span in l["spans"]:
						fsize = span["size"]
						origin = fitz.Point(span["origin"])  # the insertion point
						flags = span["flags"]
			page.addRedactAnnot(r)  # redact the word
			page.apply_redactions()
			tl = fitz.getTextlength(nombre, fontname=font, fontsize=fsize)
			# then adjust fontsize, so its fits exactly
			fsize = 2#fsize / r.width / tl#* r.width / tl
			page.insertText(origin, nombre, fontname=font, fontsize=fsize, color=blanco)

```

Thanks in advance!

**Comment by JorjMcKie** [comment]
The only way to know whether some area on the page (be it text or something else), is to walk through the page's highlight annots and then check whether your area is covered by one of the annotation rectangles.
Annotations are not part of a PDF page's contents - they are more like dust on a picture 😉.

Maybe a slightly different approach can help:
1. search for the name
2. for every hit rectangle, extract the corresponding text span and put it in some list
3. add a redact annot covering the hit rectangle (or take the span bbox for this)
4. apply redactions
5. walk through the span list, insert the replacement text and **now** add a highlight annot using the span bbox.

**Reply by JorjMcKie** [reply]
Note: If you do not insist on highlight annotations, you can also use the redact mechanism to fill the rectangle with some (permanent) color.

**Reply by mameIIas** [reply]
First of all thanks for the quick answer as always, I appreciate it a lot.
Second, that'd be a good approach if I were looking for a name indeed, I'm looking forward this:
![image](https://user-images.githubusercontent.com/34312469/161768976-a76d2e9c-d50a-40a3-b68d-5c773ba54df1.png)
but as I highlight it I get this:
![image](https://user-images.githubusercontent.com/34312469/161769082-c69ede1a-bfb5-44c1-bc46-e92c5c896bbf.png)
Is there any approach like check for a name which starts and ends with "v"?
Thanks in advance

**Reply by JorjMcKie** [reply]
Here are two snippets which do these things.
[test-annot.zip](https://github.com/pymupdf/PyMuPDF/files/8419147/test-annot.zip)

**Reply by JorjMcKie** [reply]
>  Is there any approach like check for a name which starts and ends with "v"?

Instead of using `page.search_for()` extract the page words: `page.get_text("words")`, and select those consisting of multiple letters "v". A "word" is something surrounded by spaces. It may happen, that you also get things like "vvvvv.", "vvvv-vvv", "vvvvv," etc., so your check must account for this. The returned list contains the word's bbox.

**Reply by mameIIas** [reply]
Thanks for those two snippets they give a cleaner highlight than mine!
Then for the "v" thing, an approach like this would be the goal?

`for page in doc:
    for r in page.get_text("words"):
		if r[0] == "v" and r[-1] == "v":
			
			a=page.add_highlight_annot(r)
			a.setColors({"stroke":color})
			a.update()

		`
Thanks in advance.

**Comment by JorjMcKie** [comment]
An item of `page.get_text("words")` looks like `(x0, y0, x1, y1, "word", blockn, linen, wordn)`. The first 4floats are the word bbox. The last 3 integers identify the word (numbers of block, line in block, word in line).
I would test like this:
````python
word = item[4]
lword = len(word)
bbox = fitz.Rect(item[:4])
if word == "v" * lword:  # all "v"
    # ok to process
elif lword > 2 and word[:-1] == "v" * (lword - 1):
    # ok to process: word ends with punctuation probably
else:
    # not my case
```

**Comment by JorjMcKie** [comment]
As a special service 😉, here is an example snippet:
[test-annot2.zip](https://github.com/pymupdf/PyMuPDF/files/8427039/test-annot2.zip)

**Reply by mameIIas** [reply]
Wow amazing! That was just what I was looking for, but I tested it and in spanish it does not work as good as in german. It highlights the v string just perfect, but as we have words with one letter ir also highlights them.

![image](https://user-images.githubusercontent.com/34312469/162040911-9d65e755-11dd-4072-b285-e3fce664893e.png)
 Any idea how to avoid that?
 Thanks in advance again! :)

**Reply by JorjMcKie** [reply]
I thought you are only looking for pseudo-words consisting of multiples of "v"?
Entonces como es posible selecionar un solito "a" o "y"? You do check for the length of the word, do you not?
In my example snippet I indeed left out the length check - that is probably why you are getting these letters:
`"a"[:-1]` is a string of length zero. The same is true for `"v" * 0`. So both compare equal ...
You must make sure that `len(word) > 2`.

**Comment by JorjMcKie** [comment]
More complications can arise from "words" like `(vvvvvvvv)`, `vvv-vvvvv`, etc. - all depends on the variations in your data ...
You could build a Python `set` that contains lower case "v" and any punctuation. For every word then check whether its `set(word)` is a subset of it:
```python
import string
check_set = set(string.punctuation)
check_set.add("v")
# check if word contains at least one "v" and else only punctuations:
if check_set.issuperset(set(word)) and "v" in word:
    # ok to process
```

**Comment by JorjMcKie** [comment, marked as answer]
Here are other versions that make use of that idea with Python sets - mucho más fácil!
Also added a version without annotations ...
[test-annot2.zip](https://github.com/pymupdf/PyMuPDF/files/8430281/test-annot2.zip)

**Reply by mameIIas** [reply]
Mucho más fácil indeed! Thank you so much for the answer that was really was I was looking for, just giving an amazing result:
![image](https://user-images.githubusercontent.com/34312469/162161235-c8c27a2b-5ec5-45ab-8bc6-99b06b049d95.png)
And thanks again for the help but most of all for the attention. If I have more questions we'll see each other around 😉

**Reply by JorjMcKie** [reply]
Using Python `set` is often underestimated. People tend to prematurely employ regular expressions.
In my mind they should be avoided whenever possible. Python has many, many features with strings alone.
The solution used here is able to detect all the examples I mentioned before, all sorts of punctuation anywhere.
And its use is a lot better to understand and maintain.
I at least have a hard time to understand, let alone **_debug_** a regular expression.

**Reply by mameIIas** [reply]
I agree with the regular expression, are awful to use and even to understand, another package did use it and I stopped because of that, I'll check more examples of set to get more used to it, thanks again!
